.github/workflows/publish-dockerfile.yml

```diff
@@ -0,0 +1,43 @@
+name: Build/Publish Dockerfile
+on:
+  push:
+    paths:
+      - .github/workflows/publish-dockerfile.yml
+      - docker
+  pull_request:
+    paths:
+      - .github/workflows/publish-dockerfile.yml
+      - docker
+  workflow_dispatch:
+
+
+jobs:
+  build:
+    name: Build and publish container
+    runs-on: ubuntu-latest
+    env:
+      IMAGE_NAME: cpp
+      VERSION: 0.1.0
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.PAT }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          file: ./docker/Dockerfile
+          context: ./docker
+          push: ${{ github.ref == 'refs/heads/main' }}
```

Comment:
This check is insufficient. What we need here is to check that:
1. we are on the main branch
2. there isn't already an existing container in our package repository that has the same tag, that a push would mistakenly overwrite. We don't want to overwrite an existing `cpp-gcc11-llvm-13:0.1.1` when we have forgotten to bump the version to `0.1.2`; there may not be a way to get the original back.

If both these conditions are true, it should be safe to push a new version of the container.

I'm not so experienced with GH workflow syntax, and I'm not clear on how to make this check work. Honestly, I'm not even sure if `${{ github.ref == 'refs/heads/main' }}` does what it looks like it does.

Reply:
Hmmm not familiar with GH syntax either. However, this "problem" reminds me why conan introduced revisions for recipes.  Docker images (like recipes) have different development life cycles thus you need a way to reference back every change both in code and container to have a reproducible environment. I found this post https://stackoverflow.com/questions/56212495/properly-versioning-docker-images that give some options we have about versioning that might be helpful.